### DISCUSSION TITLE
Forward Clustered Renderer Limit Prioritization

### DISCUSSION THREAD

**Discussion by mrjustaguy** [godotengine/godot-proposals#4124] [3D]
Does the FCR Use Node Process Priority to determine which Clustered elements get priority in rendering when the Limit is exceeded?

If not, How does it determine What has priority? Do Omni Lights have priority over Spot Lights etc..? Or are the priorities based on how much screen space they're taking, or is it some other method?

I'm asking as I want to set Priority to one Specific Clustered element to make sure that it Always renders, no matter what, as it is massively affecting the scene visuals (It's a Reflection Probe) while it's not a big deal if some Lights vanish if the limit is set too low.

**Comment by Calinou** [comment]
> Does the FCR Use Node Process Priority

Not to my knowledge, but you should always tweak the Max Clustered Elements setting in a way that makes your scene never exceed the limit. Otherwise, flickering is bound to happen. You can use a LOD system (such as Decal's distance fade properties) to ensure the number of clustered elements is not exceeded in the current view.

As of writing, the default value (512) is pretty high for most scenes already.

**Reply by mrjustaguy** [reply]
Well, even with 128 limit I doubt I'd exceed that limit, as I doubt I'd manage to ever have 127 lights (don't use Decals) on screen that aren't culled by OC, despite the Large draw distance of 10k that I'm using (Note: RP Renders only DL) and probably the most the player will ever be seeing is like 20-ish to 30-ish lights on screen at any given point in time.

That said, It would be nice to have the ability to give some clustered elements priority over others, as that could allow lowering the limit more while still being fairly certain the player won't be able to observe issues. I know that lowering the limit slightly improves framerate, but what I'm actually more interested is Lowering GPU Memory usage, which does seem to be affected by the limit, there's a couple hundred megabytes of difference between 128 and 8192 (the max that can be set atm) in the scene that I've been testing around..

**Reply by Calinou** [reply]
> there's a couple hundred megabytes of difference between 128 and 8192 (the max that can be set atm) in the scene that I've been testing around..

When comparing realistic units (128 vs 512 for example), it's very unlikely that the VRAM difference will be meaningful on today's GPUs, even those with only 4 GB of VRAM.

**Reply by mrjustaguy** [reply]
even 2 GB cards wouldn't see much of a difference in VRAM utilization between 128 vs 512, I mean if 128 to 8192 was only a couple hundred megabytes, 128 to 512 is like maybe a couple of 4k textures worth of VRAM.. but When the Game itself is running on a level of performance that 2 GB cards can pull off with good performance, why not cut a couple of corners that aren't even likely to ever be visible to allow them to work fine heh.
